homeassistant/helpers/entity_component.py

```diff
@@ -391,47 +357,117 @@ def async_add_entities(self, new_entities, update_before_add=False):
         if not new_entities:
             return
 
-        @asyncio.coroutine
-        def async_process_entity(new_entity):
-            """Add entities to StateMachine."""
-            new_entity.parallel_updates = self.parallel_updates
-            ret = yield from self.component.async_add_entity(
-                new_entity, self, update_before_add=update_before_add
-            )
-            if ret:
-                self.platform_entities.append(new_entity)
+        component_entities = set(entity.entity_id for entity
+                                 in self.component.entities)
 
-        tasks = [async_process_entity(entity) for entity in new_entities]
+        tasks = [
+            self._async_add_entity(entity, update_before_add,
+                                   component_entities)
+            for entity in new_entities]
 
         yield from asyncio.wait(tasks, loop=self.component.hass.loop)
         self.component.async_update_group()
 
         if self._async_unsub_polling is not None or \
            not any(entity.should_poll for entity
-                   in self.platform_entities):
+                   in self.entities.values()):
             return
 
         self._async_unsub_polling = async_track_time_interval(
             self.component.hass, self._update_entity_states, self.scan_interval
         )
 
+    @asyncio.coroutine
+    def _async_add_entity(self, entity, update_before_add, component_entities):
+        """Helper method to add an entity to the platform."""
+        if entity is None:
+            raise ValueError('Entity cannot be None')
+
+        # Do nothing if entity has already been added based on unique id.
+        if entity in self.component.entities:
+            return
+
+        entity.hass = self.component.hass
+        entity.platform = self
```

Comment:
This allows access to the platform from all platform entities but also allows the entities to change attributes of the platform. Powerful but maybe dangerous?

Reply:
They already have access to hass. But I guess you're right in that it might allow people to directly interact with entities from other platforms.

The reason I wanted to add this is to allow platforms to be able to remove entities which is something @rcloran was looking at in a recent PR.

Reply:
You could write protect some of the attributes with properties, if we think that's needed.

Reply:
This is Python, people can work around it. I'm not afraid of them assigning the platform attribute.

Reply:
Of course, we're all adults. I'm more think about mistakes. But I think it's ok like this.